python/ray/data/datasource/image_folder_datasource.py

```diff
@@ -0,0 +1,89 @@
+import pathlib
+from typing import TYPE_CHECKING, List, Optional, Union
+
+import numpy as np
+from ray.data.datasource.binary_datasource import BinaryDatasource
+from ray.data.datasource.datasource import Reader
+from ray.data.datasource.file_based_datasource import (
+    _resolve_paths_and_filesystem,
+    FileExtensionFilter,
+)
+from ray.data.datasource.partitioning import PathPartitionFilter
+
+if TYPE_CHECKING:
+    import pyarrow
+    from ray.data.block import T
+
+IMAGE_EXTENSIONS = ["png", "jpg", "jpeg", "tiff", "bmp", "gif"]
+
+
+class ImageFolderDatasource(BinaryDatasource):
+    """A datasource that lets you read datasets like `ImageNet <https://www.image-net.org/>`_."""  # noqa: E501
+
+    def create_reader(
+        self,
+        paths: Union[str, List[str]],
+        filesystem: Optional["pyarrow.fs.FileSystem"] = None,
+        partition_filter: PathPartitionFilter = None,
+        **kwargs,
+    ) -> "Reader[T]":
+        if len(paths) != 1:
+            raise ValueError(
+                "`ImageFolderDatasource` expects 1 path representing the dataset "
+                f"root, but it got {len(paths)} paths instead. To fix this "
+                "error, pass in a single-element list containing the dataset root "
+                '(for example, `paths=["s3://imagenet/train"]`)'
+            )
+
+        try:
+            import imageio  # noqa: F401
+        except ImportError:
+            raise ValueError(
```

Comment:
can you still raise ImportError here?

The problem here is that there is a missing dependency right?

Reply:
Yeah